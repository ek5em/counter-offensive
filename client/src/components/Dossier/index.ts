import Info from "./Info/Info";
import Shop from "./Shop/Shop";

export { Info, Shop };
