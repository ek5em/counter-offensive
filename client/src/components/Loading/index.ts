import CardAdvise from "./CardAdvice/CardAdvice";
import ProgressBar from "./ProgressBar/ProgressBar";

export { ProgressBar, CardAdvise };
