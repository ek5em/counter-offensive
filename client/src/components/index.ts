import Login from "./Login/Login";
import Registration from "./Registration/Registration";

export { Shop, Info } from "./Dossier";
export { ProgressBar, CardAdvise } from "./Loading";
export { Logo, Button, Input, Eye } from "./UI";
export { Login };
export { Registration };