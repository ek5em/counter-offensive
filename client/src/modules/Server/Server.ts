import { SHA256 } from "crypto-js";
import { io, Socket } from "socket.io-client";
import Mediator from "../Mediator/Mediator";
import Store from "../Store/Store";
import {
    IGamerInfo,
    IError,
    EGamerRole,
    IScene,
    IToken,
    IAnswer,
    ILobby,
    ETank,
    IMessage,
} from "./interfaces";
import { ESOCKET } from "../../config";

export default class Server {
    mediator: Mediator;
    HOST: string;
    STORE: Store;
    socket: Socket;

    constructor(HOST: string, mediator: Mediator) {
        this.HOST = HOST;
        this.mediator = mediator;
        this.STORE = new Store();

        this.socket = io(HOST);

        const {
            NEW_MESSAGE,
            LOGIN,
            SEND_MESSAGE_STATUS,
            LOGOUT,
            UPDATE_SCENE,
        } = mediator.getTriggerTypes();

        const { SERVER_ERROR, GO_TO_TANK, UPDATE_USER, LOBBY_UPDATE } =
            this.mediator.getEventTypes();

        this.socket.on(ESOCKET.ERROR, (answer: IError) => {
            mediator.call(SERVER_ERROR, answer.error);
        });

        this.socket.on(ESOCKET.GET_MESSAGE, (answer: IAnswer<IMessage[]>) => {
            mediator.get(NEW_MESSAGE, answer.data);
        });

        this.socket.on(ESOCKET.SEND_MESSAGE, (answer: IAnswer<true>) => {
            mediator.get(SEND_MESSAGE_STATUS, answer.data);
        });

        this.socket.on(ESOCKET.REGISTRATION, (answer: IAnswer<IToken>) => {
            mediator.get(LOGIN, answer.data);
        });

        this.socket.on(ESOCKET.LOGIN, (answer: IAnswer<IToken>) => {
            mediator.get(LOGIN, answer.data);
        });

        this.socket.on(ESOCKET.GET_USER_INFO, (answer: IAnswer<IGamerInfo>) => {
            this.STORE.setUser(answer.data);
            mediator.call(UPDATE_USER, answer.data);
        });

        this.socket.on(ESOCKET.LOGOUT, () => {
            mediator.get(LOGOUT);
        });

        this.socket.on(ESOCKET.TOKEN_VERIFICATION, () => {
            mediator.get(LOGIN, { token: this.STORE.getToken() });
        });

        this.socket.on(
            ESOCKET.SET_GAMER_ROLE,
            (answer: IAnswer<{ tankId: number; tankType: ETank }>) => {
                if (answer.data.tankId) {
                    mediator.call(GO_TO_TANK, answer.data);
                }
            }
        );

        this.socket.on(ESOCKET.GET_LOBBY, (answer: IAnswer<ILobby>) => {
            this.STORE.setLobby(answer.data);
            mediator.call(LOBBY_UPDATE, answer.data);
        });

        this.socket.on(ESOCKET.GET_SCENE, (answer: IAnswer<IScene>) => {
            mediator.get(UPDATE_SCENE, answer.data);
        });
    }

    // АВТОРИЗАЦИЯ

    registration(login: string, nickname: string, password: string) {
        const hash = SHA256(login + password).toString();
        this.socket.emit(ESOCKET.REGISTRATION, { login, nickname, hash });
    }

    login(login: string, password: string) {
        const rnd = Math.random();
        const hash = SHA256(
            SHA256(login + password).toString() + rnd
        ).toString();
        this.socket.emit(ESOCKET.LOGIN, { login, hash, rnd });
    }

    tokenVerification() {
        this.socket.emit(ESOCKET.TOKEN_VERIFICATION, {
            token: this.STORE.getToken(),
        });
    }

    getUser() {
        this.socket.emit(ESOCKET.GET_USER_INFO, {
            token: this.STORE.getToken(),
        });
    }

    logout() {
        this.socket.emit(ESOCKET.LOGOUT, {
            token: this.STORE.getToken(),
        });
    }

    // ЛОББИ

    sendMessage(message: string) {
        this.socket.emit(ESOCKET.SEND_MESSAGE, {
            token: this.STORE.getToken(),
            message,
        });
    }

    getMessages() {
        this.socket.emit(ESOCKET.GET_MESSAGE, {
            token: this.STORE.getToken(),
        });
    }

    setGamerRole(role: EGamerRole, tankId: number | null = null) {
        this.socket.emit(ESOCKET.SET_GAMER_ROLE, {
            token: this.STORE.getToken(),
            role,
            tankId,
        });
    }

    getLobby() {
        this.socket.emit(ESOCKET.GET_LOBBY, {
            token: this.STORE.getToken(),
        });
    }

    // ИГРА

    getScene() {
        this.socket.emit(ESOCKET.GET_SCENE, {
            token: this.STORE.getToken(),
        });
    }

    suicide() {
        this.socket.emit(ESOCKET.SUICIDE, {
            token: this.STORE.getToken(),
        });
    }

    unitMotion(x: number | null, y: number | null, angle: number) {
        this.socket.emit(ESOCKET.MOTION, {
            token: this.STORE.getToken(),
            x,
            y,
            angle,
        });
    }

    makeShot(x: number, y: number, angle: number) {
        this.socket.emit(ESOCKET.FIRE, {
            token: this.STORE.getToken(),
            x,
            y,
            angle,
        });
    }
}
