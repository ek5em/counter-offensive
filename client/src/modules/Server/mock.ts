import {
    EGamerRole,
    ERank,
    IGamerInfo,
    ILobby,
    IMessage,
    IUserUnit,
} from "./interfaces";

interface IMock {
    users: (IGamerInfo & { password: string })[];
    messages: {
        text: string;
        rank_name: ERank;
        userId: number;
        nickname: string;
    }[];
    lobby: ILobby;
    unit: IUserUnit;
}

export const MOCK: IMock = {
    users: [
        {
            gamer_exp: 9999,
            id: 0,
            is_alive: null,
            level: 10,
            login: "123",
            next_rank: 65421,
            nickname: "Гришка",
            password: "123",
            rank_name: ERank.General,
            token: "123",
        },
        {
            gamer_exp: 9999,
            id: 1,
            is_alive: null,
            level: 10,
            login: "321",
            next_rank: 65421,
            nickname: "Петька заводной",
            password: "321",
            rank_name: ERank.General,
            token: "321",
        },
    ],
    messages: [
        {
            nickname: "Гришка",
            rank_name: ERank.General,
            text: "Привет, парни!",
            userId: 0,
        },
        {
            nickname: "Петька заводной",
            rank_name: ERank.General,
            text: "Привет, Гришка, ну чо как вчера жёско зарубимся в контр-нахрюк???!",
            userId: 1,
        },
    ],
    lobby: {
        bannerman: true,
        general: true,
        tanks: {
            heavyTank: [
                { id: 1, Commander: true, Gunner: false, Mechanic: false },
                { id: 3, Commander: false, Gunner: true, Mechanic: false },
            ],
            middleTank: [{ id: 2, Gunner: true, Mechanic: false }],
        },
    },
    unit: {
        angle: 0,
        personId: EGamerRole.infantry,
        speed: 1,
        x: 5,
        y: 5,
    },
};
