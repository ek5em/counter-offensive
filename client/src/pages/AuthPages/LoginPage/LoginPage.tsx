import React, { useState } from "react";
import { Link } from "react-router-dom";
import { useGlobalContext } from "../../../hooks/useGlobalContext";
import { IUserData } from "../userData.interface";
import {
    Button,
    EButtonAppearance,
    EInput,
    Input,
    Logo,
} from "../../../components";
import { Alert } from "../Alert/Alert";

import styles from "../AuthPage.module.scss";

export const LoginPage: React.FC = () => {
    const [userData, setUserData] = useState<IUserData>({
        login: "",
        password: "",
    });
    const { server, mediator } = useGlobalContext();

    const { WARNING } = mediator.getEventTypes();

    const isValid = (login: string, password: string) => {
        if (!login || !password) {
            mediator.call(WARNING, {
                message: "Заполните все поля",
                style: "warning",
                id: "test_warning_auth_emptyFields",
            });
            return false;
        }
        return true;
    };

    const onChangeHandler = (value: string, data: string) => {
        setUserData({ ...userData, [data]: value });
    };

    const onSubmitHandler = async (e: React.FormEvent<HTMLFormElement>) => {
        e.preventDefault();
        const login = userData.login.trim();
        const pass = userData.password.trim();
        isValid(login, pass) && server.login(login, pass);
    };

    return (
        <div className={styles.authPage}>
            <Logo />
            <div className={styles.wrapper}>
                <div className={styles.content}>
                    <div className={styles.title}>
                        <Link to="/registration" tabIndex={-1}>
                            <Button
                                appearance={EButtonAppearance.primary}
                                id="test_login_goToReg_button"
                            >
                                Получить повестку
                            </Button>
                        </Link>
                    </div>
                    <form className={styles.form} onSubmit={onSubmitHandler}>
                        <div className={styles.values}>
                            <Input
                                text="Логин"
                                id="test_login_log_input"
                                value={userData.login}
                                onChange={(value) => {
                                    onChangeHandler(value, "login");
                                }}
                            />
                            <Input
                                text="Пароль"
                                id="test_login_pass_input"
                                type={EInput.password}
                                value={userData.password}
                                onChange={(value) => {
                                    onChangeHandler(value, "password");
                                }}
                            />
                        </div>
                        <Alert />
                        <Button
                            appearance={EButtonAppearance.primary}
                            className={styles.submit}
                            id="test_login_submit_button"
                        >
                            Пойти на Бахмут
                        </Button>
                    </form>
                </div>
            </div>
        </div>
    );
};
