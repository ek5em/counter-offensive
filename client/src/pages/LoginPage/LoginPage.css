.login_header {
  display: flex;
  justify-content: space-between;
  height: 20%;
  gap: 20px;
}

.login_header button {
  font-size: 60px;
  padding: 0 10px;
}

.login_wrapper {
  display: flex;
  flex-direction: column;
  height: 100%;
  width: 100%;
}

.login_content {
  display: flex;
  flex-direction: column;
  height: 100%;
  margin-top: 70px;
}

.login_submit_button {
  padding: 0 30px;
  margin: 10px 0;
  font-size: 60px;
}

.login_form {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  height: 100%;
}

.login_footer {
  display: flex;
  justify-content: flex-end;
}

.errors_div {
  max-height: 300px;
}

.errors_div > div {
	font-size: 65px;
  width: fit-content;
  font-family: Neucha;
  font-style: normal;
  line-height: normal;
  text-align: left;
  border-radius: 15px;
  border: 5px solid #000;
  max-height: 160px;
  margin: 0 0 10 0;
  padding-left: 6px;

}

.errors_div span {
  margin: 10px;
}

@media  (max-width: 1750px) {
  .errors_div > div {
    font-size: 55px;
    border: 5px solid #000;
  }
}

@media  (max-width: 1130px) {
  .errors_div > div {
	font-size: 45px;
	border: 4px solid #000;  
  }

  .login_header button,
  .login_submit_button  {
    font-size: 45px;
  }
  
}

@media  (max-width: 750px) {
  .login_header button {
    font-size: 40px;
    padding: 0 10px;
  }
  
}

@media (max-width: 550px) {
	.errors_div > div {
		font-size: 30px;
		border: 3px solid #000;
  }

  .login_submit_button{
    font-size: 40px;
    padding: 0 10px;
  }
  
}

@media (max-width: 350px) {
  .errors_div > div {
    font-size: 17px;
    border: 2px solid #000;
  }

  .login_header button {
    font-size: 30px;
  }
}
