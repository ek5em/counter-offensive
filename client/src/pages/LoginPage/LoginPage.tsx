import { FC } from "react";
import { Button, Login } from "../../components";
import "../LoginPage/LoginPage.css";
import { Link } from "react-router-dom";

const LoginPage: FC = () => {
  return (
    <div className="login_wrapper">
      <div className="login_header">
        <div>
          <Button appearance="primary-disable" id="test_login_pageName_button">
            Уже служил
          </Button>
        </div>
        <div>
          <Link to="/registration">
            <Button appearance="primary" id="test_login_goToReg_button">
              Получить повестку
            </Button>
          </Link>
        </div>
      </div>
      <div className="login_content">
        <Login />
      </div>
    </div>
  );
};

export default LoginPage;
