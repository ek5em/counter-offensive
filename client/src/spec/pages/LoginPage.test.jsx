import { screen, render, fireEvent } from "@testing-library/react";
import { LoginPage } from "../../pages/AuthPages/LoginPage/LoginPage";
import { MOCK } from "../../modules/Server/mock";
import { ContextProvider } from "../../components";

describe("Страничка авторизации", () => {
    const getLoginInput = () => screen.getByPlaceholderText(/логин/i);
    const getPasswordInput = () => screen.getByPlaceholderText(/пароль/i);
    const getSubmitBtn = () =>
        screen.getByRole("button", { name: /пойти на бахмут/i });
    const getNavBtn = () =>
        screen.getByRole("button", { name: /получить повестку/i });

    const createFields = ({
        login = "11111111",
        password = "11111111",
    } = {}) => {
        render(
            <ContextProvider useMock={true}>
                <LoginPage />
            </ContextProvider>
        );

        fireEvent.change(getLoginInput(), { target: { value: login } });
        fireEvent.change(getPasswordInput(), { target: { value: password } });
        fireEvent.click(getSubmitBtn());
    };

    test("Наличие всех элементов", () => {
        const {asFragment} = render(
            <ContextProvider>
                <LoginPage />
            </ContextProvider>
        );
        expect(getLoginInput()).toBeInTheDocument();
        expect(getPasswordInput()).toBeInTheDocument();
        expect(getSubmitBtn()).toBeInTheDocument();
        expect(getNavBtn()).toBeInTheDocument();

        expect(asFragment()).toMatchSnapshot();
    });

    test("Пустой логин", () => {
        createFields({ login: "" });
        expect(screen.getByText(/Заполните все поля/i)).toBeInTheDocument();
    });

    test("Пустой пароль", () => {
        createFields({ password: "" });
        expect(screen.getByText(/Заполните все поля/i)).toBeInTheDocument();
    });

    test("Пустой пароль и логин", () => {
        createFields({ password: "", login: "" });
        expect(screen.getByText(/Заполните все поля/i)).toBeInTheDocument();
    });

    test("Неверный логин", () => {
        createFields();
        expect(
            screen.getByText(/Неверный логин или пароль/i)
        ).toBeInTheDocument();
    });

    test("Верные данные", () => {
        createFields({ ...MOCK.users[0] });
        expect(
            screen.queryByText(/Неверный логин или пароль/i)
        ).not.toBeInTheDocument();

        expect(
            screen.queryByText(/Заполните все поля/i)
        ).not.toBeInTheDocument();
    });
});
