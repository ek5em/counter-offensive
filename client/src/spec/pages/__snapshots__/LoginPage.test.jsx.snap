// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Страничка авторизации Наличие всех элементов 1`] = `
<DocumentFragment>
  <div
    class="authPage"
  >
    <div
      class="logo"
    >
      <a
        href="/"
      >
        <img
          alt="counter-nastup"
          src="Logo.png"
        />
      </a>
    </div>
    <div
      class="wrapper"
    >
      <div
        class="content"
      >
        <div
          class="title"
        >
          <a
            href="/registration"
            tabindex="-1"
          >
            <button
              class="button primary"
              id="test_login_goToReg_button"
            >
              Получить повестку
            </button>
          </a>
        </div>
        <form
          class="form"
        >
          <div
            class="values"
          >
            <div
              class="wrapper"
            >
              <input
                autocomplete="off"
                class="input"
                id="test_login_log_input"
                placeholder="Логин"
                type="text"
              />
            </div>
            <div
              class="wrapper"
            >
              <input
                autocomplete="off"
                class="input password"
                id="test_login_pass_input"
                placeholder="Пароль"
                type="password"
              />
              <div
                class="eye"
              >
                <svg
                  data-testid="close-eye-icon"
                >
                  closeEye.svg
                </svg>
              </div>
            </div>
          </div>
          <div
            class="alert disabled"
            id=""
          >
            <span />
          </div>
          <button
            class="button submit primary"
            id="test_login_submit_button"
          >
            Пойти на Бахмут
          </button>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
